### PR TITLE
More use of sympy internally

### PR DESCRIPTION
**Pull request by lbello**
This is part of the effort to extend the use of sympy internally. See #158 for details

### CODE REVIEW

**Comment by adjs** [comment]
I was using this version in a project and one of the executions in ibmqx4 I received the following message

qiskit._resulterror.ResultError: QASM_NOT_VALID: Error parsing QASM. Error parsing qasm number 0. Parse error on line 5:
...cr[1];u2(pi,-2*atan(tan(0.75*pi))) q[1]

File: qiskit/mapper/_mapping.py [hunk 1]
```diff
@@ -692,8 +699,6 @@ def optimize_1q_gates(circuit):
                                               right_parameters[0],
                                               right_parameters[1],
                                               right_parameters[2])
-                # Evaluate the symbolic expressions for efficiency
-                right_parameters = tuple(map(sympy.N, list(right_parameters)))
```
Comment (awcross1) [inline qiskit/mapper/_mapping.py]: If we simplify a long sequence of u3 gates, is there concern that the expression size will grow? We can test this.

File: qiskit/unroll/_jsonbackend.py [hunk 2]
```diff
@@ -144,9 +144,9 @@ def u(self, arg, qubit, nested_scope=None):
             qubit_indices = [self._qubit_order_internal.get(qubit)]
             self.circuit['operations'].append({
                 'name': "U",
-                'params': [arg[0].real(nested_scope),
-                           arg[1].real(nested_scope),
-                           arg[2].real(nested_scope)],
+                'params': [arg[0].sym(nested_scope),
```
Comment (awcross1) [inline qiskit/unroll/_jsonbackend.py]: The comment at the top of the file needs to be updated to be consistent.

Comment (awcross1) [inline qiskit/unroll/_jsonbackend.py]: Actually I have a doubt about whether these should remain floats or become syms until we verify backend behavior. Perhaps related to the example @adjs gave in the conversation for this PR?

**Comment by awcross1** [comment]
These commits address the issue @adjs raised. Thank you @adjs! I also took the opportunity to change the latex() methods to use sympy. Once all of the backends accept the qobj format, the expressions in "compiled_circuit_qasm" in the qobj can stay in symbolic form.

**Comment by awcross1** [comment]
This last set of commits 1) corrected a few issues that arise for some common examples, mainly issues with using is_zero 2) added a new mapper test that highlights the problem and 3) opted to call evalf() prior to composing u3 operators due to the massive difference in time between symbolic and numeric solution. We can aim to cleanly separate approximate and exact rewriting rules as we go forward.

I think the PR is ready for review and possible merging.

**Comment by diego-plan9** [comment]
Thanks everyone involved in this large PR - it's looking ready for merging indeed!

**Pull request merged**